File: is-graph-bipartite/is-graph-bipartite.go
import ("fmt")
​
func isBipartite(graph [][]int) bool {
    color := make(map[int]int)
    
    for i := range graph {
        if _, found := color[i]; !found {
            stack := []int{i}
            color[i] = 0
            for len(stack) > 0 {
                node := stack[len(stack)-1]
                stack = stack[:len(stack)-1]
                for _, neighbor := range graph[node] {
                    if _, found := color[neighbor]; !found {
                        stack = append(stack, neighbor)
                        color[neighbor] = color[node] ^ 1
                    } else if color[neighbor] == color[node] {
                        return false
                    }
                }
            }
        }
    }
    return true
}

File: is-graph-bipartite/is-graph-bipartite.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        
        seen = {}
        stack = []
        for i in range(len(graph)):
            if i not in seen:
                stack.append(i)
                seen[i] = 1
                while stack:
                    node = stack.pop()
                    
                    for neighbor in graph[node]:
                        if neighbor in seen:
                            if not seen[node]^seen[neighbor]:
                                return False
                        else:
                            stack.append(neighbor)
                            seen[neighbor] = not seen[node]
        
        return True